### ISSUE TITLE
folding template strings

### ISSUE THREAD

**Issue by Andersgee** [moalamri/vscode-inline-fold#20]
The default regex will not handle things like
```js
<div className={`p-1 ${big ? "mt-6" : "mt-0"}`}>
```

Anyway to handle this nicely?

I tried combining a "string regex" with a "template string" regex unforunately that puts the relevant result of the second regex in group 5 instead of group 2 as the first.

[see my attempt here](https://regex101.com/r/xH3IOL/1)

**Comment by moalamri** [comment]
Try this:

https://regex101.com/r/8aHyc0/1

**Comment by moalamri** [comment]
> Try this:
> 
> https://regex101.com/r/8aHyc0/1
> 
> The capture group is 2

It doesn't fully work 

https://regex101.com/r/drLGA2/1

Hmm, maybe a regex expert could help you? I will try to make it, but I don't promise xD

**Comment by Andersgee** [comment]
Thanks for the help. I will use that for now, although it incorrectly folds everything after className, for example here
```js
<button className={`p-2 ${big ? "mt-2" : "mt-0"} mb-5`} disabled="true">
```
it results in 
```
<button className={...}>
```

**Comment by moalamri** [comment]
Hey @Andersgee 
Sorry I missed that part 💢 
I really can't think of a regex for such situation. Maybe there's a better way, but here's what I have came up with:
https://regex101.com/r/0PRtOB/1

Keep in mind that it will also fold the enclosers.

**Comment by franzwilhelm** [comment]
@moalamri try? The regex highlights the correct part, but it needs to be implemented in order for it to work in vscode

**Comment by moalamri** [comment]
> @moalamri try? The regex highlights the correct part, but it needs to be implemented in order for it to work in vscode

Go to the extension settings and set the regex to

`(className)=({`(?:.+)`}|"(?:[^`{}=\"]+)\")`

**Comment by moalamri** [comment]
@Andersgee 
Can I mark this as solved?

**Comment by franzwilhelm** [comment]
@moalamri it should be added to the code of the repo, not something you should need to configure

**Comment by moalamri** [comment]
Hey @franzwilhelm 
Why limit the extension to a single use case when it can literally do more than just folding tsx classes? Another use case I found is users use the extension to beatify markdown urls. Should I create another extension to support markdown urls folding or just give the current extension the ability to extend to the user requirements? :D

**Comment by Andersgee** [comment]
@moalamri 
yes  ```(className)=({`(?:.+)`}|"(?:[^`{}="]+)")``` works

just a note: I expect this issue to come up again unless its set as default. Its apparently not a trivial regex and I would guess close to 100% of users downloading the extension wants it to behave this way.

Thanks

**Comment by moalamri** [comment]
@franzwilhelm @Andersgee 
The problem with the regex I came up with to solve this is it instead of only capturing the matched group, it will also capture the enclosures.
Meaning the className attribute will look like this `className=...` while it should look like this `className="..."`. If someone could come up with a regex that could make the regex look as what in the screenshot I will definitely make it the extension's default regex.
What would even be better is to give examples of use cases in the extension's README.md
I will leave that to users to submit PRs to add regexs or custom settings they used to a specific case.

**Comment by moalamri** [comment]
@Andersgee @franzwilhelm 
I think I have found a workaround for this.
I would appreciate if you could help me by providing examples of classes templates so I can publish the update with this being solved

Thank you

**Comment by JasperSurmont** [comment]
I came up with a RegEx that is not perfect yet, but is going in the right direction I believe:
https://regex101.com/r/dSbZeZ/1
It still fails if you have multiple braces `{}` in your code. However, it should work fine when you simply do things like
```
class=`blabla ${test} blabla`
```

**Comment by e-medeiros** [comment]
@moalamr
 I made this regex.
``(className)=((["'])|{`)(.+?)((\2)(?!{`)|`})``
capture group is 4; 

it got all cases from @Andersgee.
but the extension rendering got messy.

**Comment by moalamri** [comment]
> @moalamri I made this regex. `` (className)=((["'])|{`)(.+?)((\2)(?!{`)|`}) `` capture group is 4;
> 
> it got all cases from @Andersgee. but the extension rendering got messy.

Hi, Would you kindly try the regex with the previous version of the extension?
You can select the previous version from the extension page.

**Comment by e-medeiros** [comment]
@moalamri worked fine with 0.1.5

**Comment by moalamri** [comment]
Would you try to see if this issue is present #24

**Comment by e-medeiros** [comment]
@moalamri still present :(

**Comment by moalamri** [comment]
@e-medeiros thanks for testing :)
The problem with react syntax is that it can have any attribute value as there's no compiler to parse the value. It's just the browser engine that renders the value as it is, and you can basically concat anything. In this case you will have to make the regex greedy, but that means any following attribute will be matched as well. When you make the regex lazy then it will stop when it encounter any of ending characters.

I will have some time next week to look into away to workaround this.

**Comment by Andersgee** [comment]
I feel like allowing setting multiple capture groups is the way to go, I looked at the source and it does not seem like a huge change?

**Comment by moalamri** [comment]
@Andersgee Yes that could work. I will look into it.

**Comment by moalamri** [comment]
Ho @Andersgee 
Thank you for assisting me in this case.
I wanted to ask if only className value can have template string. Or you can also have another attribute value in the same line that has concatenation? Let's say for example `style` or `id` values?

**Comment by Andersgee** [comment]
template strings can be anywhere (it is just normal javascript which can be anywhere).

**Comment by moalamri** [comment]
It's more like js in html in js 🤔
It's getting complicated, but I'm working on regex-free solution and I hope I somehow manage to get this done.

**Comment by moalamri** [comment]
Hi
The only method I have found to be functional in template strings is to make the regex greedy and add an extra check if there's any captured attribute then the attribute with its value are removed to have only the first captured attribute.
I found it difficult because the extension is dealing with raw text rather than rendered, so there's no way of telling the extension where to stop unless I check for any extra captured text and remove them. I hope this will be the ultimate solution to this. 
@Andersgee Thanks for your effort
